### PR TITLE
feat(schema): Schema validation at build time

### LINKED ISSUE #259: Validate scenarios against schema at build time
Currently the JSON Schema validation is done against files in `src/data/*.json` (`npm run json:check`), but it makes more sense to integrate `ajv` into the application, and render the scenarios that are imported against the schema at build time0

### PR DESCRIPTION
**Pull request by AlexAxthelm**
Introduces a "live" validation system for scenario JSON data files which runs as part of the build process via a plugin, but is also integrated into testing suite, and as `npm run json:check`. When running `npm run dev` with a non-validating dataset, there's an application crash, with the same error in the browser devtools. Any non-validating data in `src/data` will get caught at some point, probably early.

----
**Schema validation and data loading improvements:**

* Replaced manual imports in `scenariosData.ts` with dynamic loading of all JSON files in `src/data` using `import.meta.glob`, and now validates them at load time via the new `validateScenarios` utility. (`src/data/scenariosData.ts`)
* Added a new utility, `validateScenarios`, which uses AJV and the project’s JSON schema to validate scenario data, aggregating errors for better developer feedback. (`src/utils/validateScenarios.ts`)
* Added a comprehensive test suite to verify scenario data and schema enforcement, covering both real data and edge cases. (`src/utils/validateScenarios.test.tsx`)

**Build and script integration:**

* Integrated schema validation into the Vite build process with a custom plugin, so builds fail on invalid data; also serves the schema via dev server and copies it to the output. (`vite.config.ts`) [[1]](diffhunk://#diff-6a3b01ba97829c9566ef2d8dc466ffcffb4bdac08706d3d6319e42e0aa6890ddL1-R16) [[2]](diffhunk://#diff-6a3b01ba97829c9566ef2d8dc466ffcffb4bdac08706d3d6319e42e0aa6890ddR125-R177)
* Replaced the `json:check` script to use a new TypeScript script (`scripts/schema-check-files.ts`) for schema validation, and added supporting dependencies (`ajv`, `ajv-formats`, `ts-node`, `vite-plugin-static-copy`). (`package.json`, `scripts/schema-check-files.ts`, `tsconfig.scripts.json`) [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L17-R29) [[2]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R52) [[3]](diffhunk://#diff-be67053f82d58bf54c2267d4753dd88c1b458a8ea53ca0f7938124d9e4cce47bR1-R24) [[4]](diffhunk://#diff-d93b3cf7f080e09b66a979ad61f151d313c6dda8a255ef88268a361a04648597R1-R10)

**ESLint and configuration updates:**

* Updated ESLint config to apply stricter rules for source files and more relaxed rules for scripts, improving linting accuracy for different parts of the codebase. (`eslint.config.ts`) [[1]](diffhunk://#diff-3202509a51dddde1968b20a04dfcedc90615eb6681cca0cda5f43a418a516857L12-R12) [[2]](diffhunk://#diff-3202509a51dddde1968b20a04dfcedc90615eb6681cca0cda5f43a418a516857R25-R42)add schema validation and failing schema to sample data

### CODE REVIEW

File: vite.config.ts [hunk 1]
```diff
@@ -110,11 +122,55 @@ function buildInfoPlugin(): Plugin {
...
+  };
+}
+
 // https://vitejs.dev/config/
 export default defineConfig({
-  plugins: [react(), buildInfoPlugin()],
+  plugins: [
+    react(),
+    buildInfoPlugin(),
+    dataValidationPlugin("src/data"),
```
[the remark this answers is not in the document]
Comment (AlexAxthelm) [reply vite.config.ts:155]: resolved in https://github.com/RMI/pbtar/pull/260/commits/490df600916607f145ef2cf87670dd2d3d133b5c

File: vite.config.ts [hunk 2]
```diff
@@ -110,11 +122,55 @@ function buildInfoPlugin(): Plugin {
   };
 }
 
+function dataValidationPlugin() {
+  return {
+    name: "data-validation",
+    apply: "build",
+    enforce: "pre",
+    async buildStart() {
+      const dir = "src/data";
```
[the remark this answers is not in the document]
Comment (AlexAxthelm) [reply vite.config.ts]: Resolved in https://github.com/RMI/pbtar/pull/260/commits/490df600916607f145ef2cf87670dd2d3d133b5c

File: src/utils/validateScenarios.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        .map((e) => `${e.instancePath || "/"} ${e.message}`)
+        .join("\n  ");
+      problems.push(`✖ ${name}\n  ${msgs}`);
+      continue;
+    }
+    if (!Array.isArray(data)) {
+      problems.push(`✖ ${name}\n  expected an array of Scenario items`);
+      continue;
+    }
+    out.push(...(data as Scenario[]));
```
[the remark this answers is not in the document]
Comment (AlexAxthelm) [reply src/utils/validateScenarios.ts]: Resolved in https://github.com/RMI/pbtar/pull/260/commits/e0a5c2495d3243545f834fccead57417ce4d56e3

File: src/utils/validateScenarios.test.tsx [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+import { Scenario } from "../types";
+
+describe("real scenario data validity", () => {
+  it("all src/data/*.json data files conform to schema", async () => {
+    const dir = resolve(__dirname, "../data");
+    const names = (await fs.readdir(dir)).filter((f) => f.endsWith(".json"));
+    const entries: FileEntry[] = await Promise.all(
+      names.map(async (name) => ({
+        name,
+        data: JSON.parse(await fs.readFile(join(dir, name), "utf8")) as string,
```
[the remark this answers is not in the document]
Comment (AlexAxthelm) [reply src/utils/validateScenarios.test.tsx]: Resolved in https://github.com/RMI/pbtar/pull/260/commits/e0a5c2495d3243545f834fccead57417ce4d56e3

File: src/data/scenariosData.ts [hunk 5]
```diff
@@ -1,23 +1,18 @@
...
-  ...scenarios_metadata_minimal,
-];
+const entries: FileEntry[] = Object.entries(modules)
+  .map(([path, data]) => ({
+    name: path.split("/").pop()!, // e.g. "scenarios_metadata_1.json"
+    data, // file contents
+  }))
+  .sort((a, b) => a.name.localeCompare(b.name));
+
+export const scenariosData: Scenario[] = validateScenarios(entries).flat();
```
[the remark this answers is not in the document]
Comment (AlexAxthelm) [reply src/data/scenariosData.ts]: Resolved in https://github.com/RMI/pbtar/pull/260/commits/e0a5c2495d3243545f834fccead57417ce4d56e3

File: scripts/generate-schema-docs.sh [hunk 6]
```diff
@@ -6,10 +6,13 @@ python3 -m venv ".venv/python"
 # This adds the generate-schema-doc command
 pip3 install "json-schema-for-humans"
 
+# Create djirectory if no exist, and `cd` to it
```
Comment (jdhoffa) [inline scripts/generate-schema-docs.sh]:
```suggestion
# Create directory if no exist, and `cd` to it
```

File: src/data/scenariosData.ts [hunk 7]
```diff
@@ -1,23 +1,18 @@
...
-  ...scenarios_metadata_minimal,
-];
+const entries: FileEntry[] = Object.entries(modules)
+  .map(([path, data]) => ({
+    name: path.split("/").pop()!, // e.g. "scenarios_metadata_1.json"
+    data, // file contents
+  }))
+  .sort((a, b) => a.name.localeCompare(b.name));
+
+export const scenariosData: Scenario[] = validateScenarios(entries);
```
Comment (jdhoffa) [inline src/data/scenariosData.ts:18]: <3

**Review by jdhoffa — DISMISSED** [review]
I'm honestly not qualified to properly review a lot of the `.ts` changes here. I've given a cursory look to check that I understand what is going on, but I don't have enough experience to catch any gotchas or anti-patterns. 

I've tested functionality locally, by removing a required field:
Screenshot 2025-08-25 at 12 38 01

and confirm that `npm run json:check` gives helpful info:
Screenshot 2025-08-25 at 12 38 38

also confirm that running `npm run build` gets me a build-time failure, which is awesome:
Screenshot 2025-08-25 at 12 40 12

Love the comprehensive tests!!

**Comment by jdhoffa** [comment]
@AlexAxthelm feel free to re-request review once merge conflicts are resolved

**Comment by jdhoffa** [comment]
@AlexAxthelm I think schema docs need to be re-generated

**Review by jdhoffa — APPROVED** [review]
